cmd/gateway-gcs.go

```diff
@@ -268,17 +279,67 @@ func newGCSGateway(projectID string) (GatewayLayer, error) {
 		return nil, err
 	}
 
-	return &gcsGateway{
-		client:     client,
-		projectID:  projectID,
-		ctx:        ctx,
-		anonClient: anonClient,
-	}, nil
+	gateway := &gcsGateway{
+		client:       client,
+		projectID:    projectID,
+		ctx:          ctx,
+		anonClient:   anonClient,
+		cleanupEndCh: make(chan struct{}),
+	}
+	// Start background process to cleanup old files in minio.sys.temp
+	go gateway.CleanupGCSMinioPath()
+	return gateway, nil
+}
+
+// Cleanup old files in minio.sys.temp of the given bucket.
+func (l *gcsGateway) CleanupGCSMinioPathBucket(bucket string) {
+	it := l.client.Bucket(bucket).Objects(l.ctx, &storage.Query{Prefix: gcsMinioPath, Versions: false})
+	for {
+		attrs, err := it.Next()
+		if err != nil {
+			if err != iterator.Done {
+				errorIf(err, "Object listing error on bucket %s", bucket)
+			}
+			return
+		}
+		if time.Since(attrs.Updated) > gcsMultipartExpiry {
+			// Delete files older than 1 week.
+			err := l.client.Bucket(bucket).Object(attrs.Name).Delete(l.ctx)
+			if err != nil {
+				errorIf(err, "Unable to delete the object %", attrs.Name)
+				return
+			}
+		}
+	}
+}
+
+// Cleanup old files in minio.sys.temp of all buckets.
+func (l *gcsGateway) CleanupGCSMinioPath() {
+	for {
+		it := l.client.Buckets(l.ctx, l.projectID)
+		for {
+			attrs, err := it.Next()
+			if err != nil {
+				if err != iterator.Done {
+					errorIf(err, "Bucket listing error")
+				}
+				break
+			}
+			l.CleanupGCSMinioPathBucket(attrs.Name)
+		}
+		select {
+		case <-l.cleanupEndCh:
+			// If shutdown is called we end the cleanup process.
+			return
+		case <-time.After(gcsCleanupInterval):
+		}
+	}
 }
 
 // Shutdown - save any gateway metadata to disk
 // if necessary and reload upon next restart.
 func (l *gcsGateway) Shutdown() error {
+	close(l.cleanupEndCh)
```

Comment:
The purpose here seems to be to make the cleanup go-routine quit, however, if it is in the middle of scanning buckets it will fail to quit immediately, and in the meanwhile the rest of the process will quit. So it does not seem to be needed.